teps/0048-pipeline-as-code.md

```diff
@@ -0,0 +1,214 @@
+- [Upgrade &amp; Migration Strategy (optional)](#upgrade--migration-strategy-optional)
+- [References (optional)](#references-optional)
+
+## Summary
+
+The pipelines as code technique emphasizes that the configuration of delivering
+pipelines that build, test and deploy our applications or infrastructure should
+be treated as code; they should be placed under source control and modularized
+in reusable components with automated testing and deployment.
+
+
+## Motivation
+
+The Tekton pipeline definitions are currently dissociated from a source code repository. 
+They may be located in the same repository as they code they should build, test and deploy,
+but Tekton is not aware of them. Tekton expects the pipeline definition to be deployed to a
+kubernetes cluster. Even when the Tekton resources are deployed to the cluster from the
+Repository, it's not easy to test changes to them, as their definition won't be updated until
+they are redeployed to the cluster. 
+
+The reason for the current behaviour is that Tekton is non-prescriptive about how to set up
+CI and CD pipelines. It aims to provide the building blocks for creating such pipelines and it leaves
+it up to the user to decide how to integrate the pipeline in the overall workflow. 
+
+The use case of a single repo, running CI/CD jobs in isolation, is very common and
+we wish to provide dedicated support for it, by providing an easy way for users to
+run pipelines and tasks stored in a specific folder of the repo. The same tool can be used by
+pipeline and task authors for quick testing and experimentation of their code.
+
+### Goals
+
+A way for users to add a tekton pipeline in a repository and getting it
+automatically picked up for CI or CD.
+
+> The solution should be as much as possible git hosting system agnostic.
+
+### Non-Goals
+
+- GitOps: Pipelines as Code does not take care of the [GitOps primitives](https://www.weave.works/technologies/gitops/), it
+  simply applies tekton pipelines that may take care of some GitOps operation.
+
+### Use Cases
+
+- A User have a source code repository based on some Golang backend.
+- User add a webhook (or other means) from its repository to pipelines as a code.
+- The user add a directory called in `.tekton` with some pipeline definition for
+  testing its repository.
+- User sending a PR with its change.
+- The source code is automatically tested with the pipeline definition in `.tekton`.
+
+Subsequently if the user needs to add another component to its source code (i.e:
+a javascript based frontend) and need to make a changes to the pipeline
+definition to do the testing of that new component, the change to the pipeline
+would only apply for that Pull Request until it will be merged.
+
+## Requirements
+
+* We need to be able to test changes inside the `.tekton` folder as part of the PR sent.
+* We should provide an easy way (single curl command and/or operator flag) to deploy Tekton with "tekton-as-code" support.
```

Comment:
i dont exactly disagree with this requirement, but it seems to be specific to a particular solution (e.g. if we solved this by building support into triggers or pipelines, this might not be relevant?)

Reply:
That's true *but* work on "pipeline as code" can start / take place in parallel of getting this in triggers (or pipeline 😓 ). I think it's a requirement to "bootstrap" the work on the topic (giving time to implement this correctly if need be in tekton components).

Reply:
I agree with @vdemeester 

I don't think we need to make "triggering as code" a requirement for pipeline as code.

Reply:
> I don't think we need to make "triggering as code" a requirement for pipeline as code.

can we expand on this a bit? i DO think that we should aim for this - @chmouel you pointed out some other solutions that do not allow triggering as code, but i strongly feel that this is where we want to get to (solutions that don't currently provide it i'm guessing would if they could - and probably will eventually)

maybe one way to avoid expanding the scope of what you are planning too far is if we could include a requirement that for whatever solutions we look at, we at least consider how they could be expanded to support triggering as code? <-- this depends on whether we want what we explore here to be the tekton recommended way of doing pipelines as code, or whether we want this to be one of many options that tekton provides. if this is "the one tekton way" then i think we need to be a lot more thorough